chef/automate
# [automate-854] Project update status on Projects List page

### :nut_and_bolt: Description: What code changed, and why?

This first timing diagram shows nominal behavior for adding/editing rules for a project. The user at the top initiates various actions. The two sections below, "Ingest Rules" and "Project Update Status" represent the column names on the Projects List page.

This PR deals mostly with populating the "Project Update Status"  column with live data; "Ingest Rules" was done in an earlier PR, though it is fine-tuned here.
This PR includes front-end support for the "Cancel" operation; however, it is not yet supported by the back-end. Jay's imminent work should close the loop on that.

![nominal](../raw/ms/project-update-status/dev-docs/diagrams/rule-updates-timing-nominal.png)

Similarly, this shows the sequence around a user cancelling an update:
![cancelling](../raw/ms/project-update-status/dev-docs/diagrams/rule-updates-timing-cancel.png)

In order to populate those columns on the Projects List page, this adaptive sequence of polling the rules status is applied:

![adaptive timing](../raw/ms/project-update-status/dev-docs/diagrams/rule-updates-status-check.png)

There are three stage of polling beginning at the moment the user confirms "Update Projects":

1. Initial Burst -- this fires 4 times a second for just 3 seconds. It provides the user some visual feedback in the event that the update process is very quick. Otherwise, they might not even see that the update started and finished.
2. Active Monitoring -- after the initial burst this fires every 5 seconds as long as the update is in process.
3. Dormant Monitoring -- as soon as the update completes, we revert to dormant monitoring, just once per minute, like leaving the pilot light on. As soon as this user--or any other user begins another update, dormant monitoring transitions again to (1) if the current user initiated the update or (2) if some other user did.

### :chains: Related Resources

PR - #1250 (Update rules status on Project List page)

### :+1: Definition of Done

Both "Ingest Rules" and "Project Update Status" columns on the Projects List page are live.


### :athletic_shoe: How to Build and Test the Change

1. rebuild automate-ui
2. In hab studio, create lots of nodes, e.g. `chef_load_nodes 10000`. That will take minutes. (You could usually see something with just 5000, to make the wait less onerous.)
3. Create project and rules.
4. Note that where you have added rules you will see "needs updating".
5. Select "Update Projects" and confirm.
➡️ Observe the progress in the Update button and soon enough the status changes to OK.
Be sure to look at the unit tests for other than happy path scenarios and for details on how the mechanism works.
